Implement exercise binary-search

chgraef opened the pull request:
Started to implement binary-search exercise.

patricksjackson commented on exercises/binary-search/example.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace binary_search {
+
+int find(std::vector<int> const&, int);
```
This should return a `size_t`, since it is returning an index.

chgraef replied:
OK, changed.

patricksjackson commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,28 @@
+            return mid; 
+        }
+    }
+        return -1;
```
```suggestion
        return SIZE_MAX;
```

chgraef replied:
The updated problem spec suggests to raise a "value not in array" error in the last 5 cases. I changed my example and tests such that an exception is thrown and tested for, respectively. What are your thoughts on this?

patricksjackson replied:
I think throwing is probably fine. I don't like it but it does fit the problem spec best :\

patricksjackson commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,28 @@
+
+int find(std::vector<int> const& data, int value)
+{
+    int lpos = 0;
```
These types are indexes so they should all be `size_t`

chgraef replied:
Thanks. Changed this to size_t throughout.

patricksjackson commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,28 @@
+
+    while (lpos <= rpos)
+    {
+        mid = (lpos + rpos)/2;
```
There is a bug on this line, by the way. Though once you change the variables to `size_t` it will effectively never be hit.

patricksjackson replied:
Sorry I should've been clear, I said this in a 'this is an interesting thing I've come across' way and not in a 'go fix this weird bug' way.

I like binary search, and I think it's a neat fact that it was first proposed in the 1940's but the first correct implementation didn't appear until the 1960's. And [most implementations are wrong due to the exact bug here](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html).

Also, I work in the optimization space, and I've used some of the ideas in [this paper](https://arxiv.org/abs/1509.05053) in interviews to explain some of the kinds of analysis work that I do. Particularly interesting are the interactions between the branchy and branch-free binary search algorithms with the cache.

chgraef replied:
Thanks for the explanations, the reading suggestions and for pointing out the bug in the first place! The first short article was a really enlightening read. Will defo put some time aside for a cup of tea and the arXiv paper tomorrow.

More of this, please! :-)

patricksjackson commented:
Copying here from what I said in #163

Obviously we (well, I) am very behind here, since this has been open almost a year and a half :\ Sorry.

I am trying to slowly clean things up around here. If you still have an interest in this, please make the suggested changes. If you're not interested any more, either respond here or do nothing by say the middle of March (2-3 weeks from now) and I'll open this up for others to finish up.

Since it's been so long there are things broken that have to do with the track itself. I am **not** asking you to fix them, since your PR was current when you made this. (You _can_ do it, but I will take on doing that myself if you like). That also applies to the test suite itself, which has had some updates - https://github.com/exercism/problem-specifications/blob/master/exercises/binary-search/canonical-data.json

patricksjackson commented on config.json:
```diff
@@ -201,6 +201,21 @@
+      "uuid": "abdd7aa3-25f6-4d27-a4dc-6bd1ea2f718c"
+    },
+    {
+      "core": false,
```
There's a merge error here where you're adding to the next problem as well. You also need to re-format your settings.

The isogram merge had the same issue, and I missed it too. (I fixed that already)

chgraef replied:
Done. Thanks for the pointer.

patricksjackson commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,40 @@
+
+    while (lpos <= rpos)
+    {
+        mid = (lpos + rpos) >> 1;
```
Technically this doesn't fix the issue. You need
```
mid = lpos + ((rpos - lpos) / 2);
```
The fix in the blog post that they cite for Cpp was using signed indices (lpos and rpos), so they could just cast them to unsigned and the addition was guaranteed to never overflow due to the extra bit. But that's assuming the indices even fit into a signed int.

(Since you did want more reading, we do use size_t here for indexes, but there is a good argument for using a signed type -> isocpp/CppCoreGuidelines#1115. )

chgraef replied:
Oh, I see. I changed the LOC accordingly.

NB: Thanks for the link, very interesting read! I understand we'd run into a narrowing conversion issue if we were to change size_t for ptrdiff_t, since vector.size() returns an unsigned type. Hope that the index type will trickle down into the standard not too far in the future...

KevinWMatthews commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,40 @@
+
+size_t find(std::vector<int> const& data, int const value)
+{
+    if(data.empty() || value < data.front() || data.back() < value)    
```
A idea on style - some prefer to extract a complex boolean into a helper function such as:
```c++
if(!is_value_in_range(data, value))
```
feeling that it adds expressiveness to the code. Others don't like the extra level of indirection. I don't know if exercism has a preference.

It works nicely as it is; just food for thought.

chgraef replied:
Good point, thanks @KevinWMatthews. I introduced a function, as per your suggestion.

KevinWMatthews commented on exercises/binary-search/example.cpp:
```diff
@@ -0,0 +1,40 @@
+            {
+                lpos = mid + 1;
+            }
+        }
```
Another thought on style - I've seen a lot of conditionals get out of control, so I can't help wanting to remove the `else`. Something like:
```c++
if (data.at(mid) == value) 
{ 
    return mid; 
}

if (value < data.at(mid)) 
{
    // snip
```
Again a matter of preference; it's perfectly fine as it is, and I'm sure some would find the current way preferable. 

Definitely not worth holding up a merge for something so small. Thanks for your implementation!

chgraef replied:
Thanks, @KevinWMatthews for your comment. This is a good point, so, I unwrapped the nested if/elses and turned these into a series of simple ifs.

chgraef commented:
Hi @patricksjackson, thanks for all your comments, which are now addressed. Please feel free to merge if there aren't any further open points that would require attention.

KevinWMatthews commented:
Style points!

Do you like `is_value_in_range()` as part of the public interface or local to the .cpp file?

I like the `mid = lpos + ((rpos - lpos) / 2);` fix. It's a rare but real bug.

I'm sure @patricksjackson will weigh in/merge, but I'll keep an eye on this issue too.

chgraef commented:
Hello @KevinWMatthews, good point, there's no need for the helper function to be declared in the header, so I changed this and updated the pull request.

Cheers for the hint, it's much appreciated!

patricksjackson commented on exercises/binary-search/binary_search_test.cpp:
```diff
@@ -0,0 +1,114 @@
+#include <vector>
+#include <cstddef>
+
+using namespace std;
```
You can just remove this as you don't use it anywhere
```suggestion
```

patricksjackson commented:
Woo! I'll go ahead and merge.
